apm-agent-core/src/main/java/co/elastic/apm/impl/transaction/SystemClock.java

```diff
@@ -0,0 +1,74 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.impl.transaction;
+
+import org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement;
+
+import java.time.Clock;
+import java.time.Instant;
+
+public interface SystemClock {
+
+    long getEpochMicros();
+
+    enum ForCurrentVM implements SystemClock {
+        INSTANCE;
+        private final SystemClock dispatcher;
+
+        ForCurrentVM() {
+            SystemClock localDispatcher;
+            try {
+                Class.forName("java.time.Instant");
+                localDispatcher = ForJava8CapableVM.INSTANCE;
+            } catch (Exception | NoClassDefFoundError e) {
+                localDispatcher = ForLegacyVM.INSTANCE;
+            }
+            dispatcher = localDispatcher;
+        }
+
+        @Override
+        public long getEpochMicros() {
+            return dispatcher.getEpochMicros();
+        }
+    }
+
+    @IgnoreJRERequirement
+    enum ForJava8CapableVM implements SystemClock {
+        INSTANCE;
+
+        private static final Clock clock = Clock.systemUTC();
```

Comment:
You are completely right - that would be safer. But we don't support such a VM so I would not want to pay the cost of the increased complexity and potentially higher overhead yet. We can't claim to support a particular VM if we don't have an integration test for it. If we are to support such a VM in the future at that time, we might even have already dropped Java 7 support so that we can just use `Instant` right away.